src/proxy_handler.cc

```diff
@@ -0,0 +1,180 @@
+  if (client_ == nullptr) {
+    std::cerr << "ProxyHandler::HandleRequest failed to allocate AsyncClient.";
+    std::cerr << std::endl;
+    return RequestHandler::Error;
+  }
+
+  if (!client_->Connect(remote_host_, remote_port_)) {
+    std::cerr << "ProxyHandler::HandleRequest failed to connect to ";
+    std::cerr << remote_host_ << ":" << remote_port_ << std::endl;
+    return RequestHandler::Error;
+  }
+
+  if (!client_->Write(proxied_req.raw_request())) {
+    std::cerr << "ProxyHandler::HandleRequest failed to write to ";
+    std::cerr << remote_host_ << ":" << remote_port_ << std::endl;
+    return RequestHandler::Error;
+  }
+
+  std::string response_str;
+  if (!client_->Read(response_str)) {
+    std::cerr << "ProxyHandler::HandleRequest failed to read from ";
+    std::cerr << remote_host_ << ":" << remote_port_ << std::endl;
+    return RequestHandler::Error;
+  }
+
+  // copy parsed response into given response
+  *response = ParseRawResponse(response_str);
+
+  if (response->status() == Response::code_302_found ||
+      response->status() == Response::code_301_moved) {
+    Request redirect_request(request);
+    redirect_request.set_uri(redirect_uri_);
+    redirect_request.remove_header("Host");
+    redirect_request.add_header("Host", redirect_host_);
+    uri_prefix_ = "";
+    return HandleRequest(redirect_request, response);
+  }
+
+  return RequestHandler::OK;
+}
+
+// TODO: this assume raw_response is valid HTTP
+// TODO: factor this out into a more appropriate location
+Response ProxyHandler::ParseRawResponse(const std::string& raw_response)
+{
+  Response response;
+
+  // for a given substring, start points to first char, end points to 1
+  // position past the end
+  size_t start = 0, end;
+
+  // get response code
+  start = raw_response.find(" ") + 1; // skip version
+  end = raw_response.find(" ", start);
+  int raw_response_code = std::stoi(raw_response.substr(start, end - start));
+  Response::ResponseCode rc;
+  switch(raw_response_code) {
+    case 200:
+      rc = Response::code_200_OK;
+      break;
+
+    case 301:
+      rc = Response::code_301_moved;
+      break;
+    case 302:
+      rc = Response::code_302_found;
+      break;
+
+    case 400:
+      rc = Response::code_400_bad_request;
+      break;
+    case 401:
+      rc = Response::code_401_unauthorized;
+      break;
+    case 403:
+      rc = Response::code_403_forbidden;
+      break;
+    case 404:
+      rc = Response::code_404_not_found;
+      break;
+
+    case 500:
+      rc = Response::code_500_internal_error;
+      break;
+  }
+  response.SetStatus(rc);
+
+  // get all headers
+  std::string header_name, header_value;
+  end = raw_response.find("\r\n", start);
+  while (true) {
+    start = end + 2;
+
+    // if next \r\n is right after current one, we're done with the headers
+    if (raw_response.find("\r\n", start) == start) {
+      start += 2;
+      break;
+    }
+
+    end = raw_response.find(" ", start);
```

Comment:
I don't think headers have to have a space after the `:`, so this may not always work.